src/adt/edgeset.c

```diff
@@ -871,3 +887,997 @@ edge_set_ordered_iter_next(struct edge_ordered_iter *eoi, struct fsm_edge *e)
+	}
+
+	set->ceil = src->ceil;
+	memcpy(set->groups, src->groups,
+	    src->count * sizeof(src->groups[0]));
+	set->count = src->count;
+
+#if LOG_BITSET
+	{
+		size_t i;
+		for (i = 0; i < set->count; i++) {
+			fprintf(stderr, "edge_set[%zd]: to %d, [0x%lx, 0x%lx, 0x%lx, 0x%lx]\n",
+			    i, set->groups[i].to,
+			    set->groups[i].symbols[0],
+			    set->groups[i].symbols[1],
+			    set->groups[i].symbols[2],
+			    set->groups[i].symbols[3]);
+		}
+	}
+#endif
+
+	*dst = set;
+	return 1;
+}
+
+void
+edge_set_remove(struct edge_set **pset, unsigned char symbol)
+{
+	size_t i;
+	struct edge_set *set;
+
+	assert(pset != NULL);
+	set = *pset;
+
+#if LOG_BITSET
+	fprintf(stderr, " -- edge_set_remove: symbol 0x%x on %p\n",
+	    symbol, (void *)set);
+#endif
+
+	if (edge_set_empty(set)) {
+		return;
+	}
+
+	/* This does not track whether edge(s) were actually removed.
+	 * It also does not remove edge groups where none of the
+	 * symbols are set anymore. */
+	for (i = 0; i < set->count; i++) {
+		struct edge_group *eg = &set->groups[i];
+		SYMBOLS_CLEAR(eg->symbols, symbol);
+	}
+}
+
+void
+edge_set_remove_state(struct edge_set **pset, fsm_state_t state)
+{
+	size_t i;
+	struct edge_set *set;
+
+	assert(pset != NULL);
+	set = *pset;
+
+#if LOG_BITSET
+	fprintf(stderr, " -- edge_set_remove_state: state %d on %p\n",
+	    state, (void *)set);
+#endif
+
+	if (edge_set_empty(set)) {
+		return;
+	}
+
+	i = 0;
+	while (i < set->count) {
+		const struct edge_group *eg = &set->groups[i];
+		if (eg->to == state) {
+			const size_t to_mv = set->count - i;
+			memmove(&set->groups[i],
+			    &set->groups[i + 1],
+			    to_mv * sizeof(*eg));
+			set->count--;
+		} else {
+			i++;
+		}
+	}
+}
+
+struct to_info {
+	fsm_state_t old_to;
+	fsm_state_t new_to;
+	fsm_state_t assignment;
+};
+
+static int
+collate_info_by_new_to(const void *pa, const void *pb)
+{
+	const struct to_info *a = (const struct to_info *)pa;
+	const struct to_info *b = (const struct to_info *)pb;
+	if (a->new_to < b->new_to) {
+		return -1;
+	} else if (a->new_to > b->new_to) {
+		return 1;
+	} else {
+		return 0;
+	}
+}
+
+static int
+collate_info_by_old_to(const void *pa, const void *pb)
+{
+	const struct to_info *a = (const struct to_info *)pa;
+	const struct to_info *b = (const struct to_info *)pb;
+	if (a->old_to < b->old_to) {
+		return -1;
+	} else if (a->old_to > b->old_to) {
+		return 1;
+	} else {
+		assert(!"violated uniqueness invariant");
+		return 0;
+	}
+}
+
+int
+edge_set_compact(struct edge_set **pset, const struct fsm_alloc *alloc,
+    fsm_state_remap_fun *remap, const void *opaque)
+{
+	struct edge_set *set;
+	size_t i;
+	struct to_info *info;
+	struct edge_group *ngroups;
+	size_t ncount = 0;
+
+	assert(pset != NULL);
+	set = *pset;
+
+#if LOG_BITSET
+	fprintf(stderr, " -- edge_set_compact: set %p\n",
+	    (void *)set);
+#endif
+
+	if (edge_set_empty(set)) {
+		return 1;
+	}
+
+	assert(set->count > 0);
+
+	info = f_malloc(alloc, set->count * sizeof(info[0]));
```

Comment:
I feel like there should be a way to do this that doesn't need 3 words per group, though in practice we're probably coming out way ahead by collapsing together lots of different labels leading to the same state.

My main problem with this implementation is that it still does as much work when the edge set's groups are already unique and sorted, compared to something that only rearranged/compacted in place as necessary, but my other approaches got tricky.